ouroboros-consensus-test/test-consensus/Test/Consensus/Mempool/Fairness.hs

```diff
@@ -0,0 +1,207 @@
+    ----------------------------------------------------------------------------
+    let
+      nrSmall :: Double
+      nrLarge :: Double
+      (nrSmall, nrLarge) = (fromIntegral . length *** fromIntegral . length)
+                         $ List.partition  (<= smallTxSize)
+                         $ fmap txSize txs
+    length txs @?= nrOftxsToCollect
+    -- At the end of the tests the proportion of small and large
+    -- transactions that were added should be rouhgly the same. We tolerate
+    -- the given theshold.
+    let diff_ratio = abs (nrLarge - nrSmall) / (nrLarge + nrSmall)
+    diff_ratio <= toleranceThreshold
+      @? (    "The difference between the number of large and small transactions added "
+           <> "exeeds the threshold (" <> show toleranceThreshold <> ")\n"
+           <> "Total number of small transactions that were added: " <> show nrSmall <> "\n"
+           <> "Total number of large transactions that were added: " <> show nrLarge <> "\n"
+           <> "Difference / Total: " <> show diff_ratio
+         )
+
+-- | The fairness test depends on the:
+--
+-- - Total mempool capacity.
+-- - Size of what we consider to be a small transaction.
+-- - Size of what we consider to be a large transaction.
+-- - How many added transactions we count.
+--
+-- The smaller the difference between 'smallTxSize' and 'largeTxSize', the
+-- harder it will be detect potential differences in way the mempool handles
+-- small and large transactions.
+--
+-- The larger the capacity, the higher the chance large transactions can be
+-- added before the mempool is saturated.
+--
+-- We tolerate a certain ratio between the difference of small and large
+-- transactions added, and the total transactions that were added. For instance,
+-- given a threshold of 0.2, if we measure the sizes of 100 added transactions,
+-- the difference between the number small and large transactions we counted
+-- should not be larger than 20.
+--
+data TestParams = TestParams {
+    mempoolMaxCapacity :: Word32
+  , smallTxSize        :: Word32
+  , largeTxSize        :: Word32
+  , nrOftxsToCollect   :: Int
+  , toleranceThreshold :: Double
+  }
+
+-- | Add transactions with with the specified size to the mempool.
+--
+-- We launch a fixed number of adder threads, set to 2 at the moment.
+--
+-- This process does not finish. Hence the 'a' type parameter.
+adders ::
+     TestMempool
+     -- ^ Mempool to which transactions will be added
+  -> Word32
+     -- ^ Transaction size
+  -> IO a
+adders mempool fixedTxSize = Async.mapConcurrently_ adder [0..2] >> error "This process should not finish!"
+  where
+    adder :: Int -> IO ()
+    adder _i = forever $ do
+        -- We pick a random number for the transaction id.
+        thisTxId <- randomIO
+        void $ Mempool.addTxs mempool [mkGenTx thisTxId fixedTxSize]
+
+-- | Remove the given number of transactions and return them.
+--
+-- POSTCONDITION: pure (Just n) = fmap (fmap lenght) $ remover mempool n
```

Comment:
I like the idea of making explicit post-conditions but then why not put them in the code as `assert`? This will ensure they stay current and the code is correct (here, there's a typo :) )